src/main/java/com/google/devtools/build/lib/skyframe/toolchains/ToolchainResolutionDebugPrinter.java

```diff
@@ -13,43 +13,69 @@
 // limitations under the License.
 package com.google.devtools.build.lib.skyframe.toolchains;
 
+import static com.google.common.base.Preconditions.checkNotNull;
 import static java.util.stream.Collectors.joining;
 
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.ImmutableSetMultimap;
+import com.google.devtools.build.lib.analysis.config.ToolchainTypeRequirement;
 import com.google.devtools.build.lib.analysis.platform.ConstraintValueInfo;
 import com.google.devtools.build.lib.analysis.platform.ToolchainTypeInfo;
 import com.google.devtools.build.lib.cmdline.Label;
+import com.google.devtools.build.lib.cmdline.RepositoryMapping;
+import com.google.devtools.build.lib.cmdline.RepositoryName;
 import com.google.devtools.build.lib.events.Event;
+import com.google.devtools.build.lib.events.EventHandler;
 import com.google.devtools.build.lib.events.ExtendedEventHandler;
+import com.google.devtools.build.lib.skyframe.RepositoryMappingValue;
+import com.google.devtools.build.skyframe.SkyFunction;
 import com.google.errorprone.annotations.FormatMethod;
 import com.google.errorprone.annotations.FormatString;
+import java.util.HashSet;
 import java.util.Map;
 
 /** A helper interface for printing debug messages from toolchain resolution. */
 public sealed interface ToolchainResolutionDebugPrinter {
 
-  static ToolchainResolutionDebugPrinter create(boolean debug, ExtendedEventHandler eventHandler) {
-    if (debug) {
-      return new EventHandlerImpl(eventHandler);
+  static ToolchainResolutionDebugPrinter create(boolean debug, SkyFunction.Environment env)
+      throws InterruptedException {
+    if (!debug) {
+      return new NoopPrinter();
     }
-    return new NoopPrinter();
+    var mainRepositoryMapping =
+        (RepositoryMappingValue) env.getValue(RepositoryMappingValue.key(RepositoryName.MAIN));
+    checkNotNull(
+        mainRepositoryMapping,
+        "expected main repository mapping to have been computed when resolving toolchains");
+    return env.getState(() -> new EventHandlerImpl(mainRepositoryMapping.repositoryMapping()));
   }
 
   boolean debugEnabled();
 
-  /** Report on which toolchains were selected. */
-  void reportSelectedToolchains(
+  void startDebugging(
+      EventHandler eventHandler,
```

Comment:
Since the `eventHandler` ultimately comes from the `SkyFunction.Environment`, why not save it during the `create` call? Why pass it to every method? Are the instances changing?

Reply:
Yes, there is a check in Skyframe that specifically guards against this.